### ISSUE TITLE
Include library and keyword source information in Libdoc spec files

### ISSUE THREAD

**Issue by fabioz** [robotframework/robotframework#3507]
Provide which are the sources that generated a given .libspec file (when possible).

For instance, if a keyword was generated from a method in some Python file, I'd like to track that file for modifications to know that I have to regenerate the spec if that file changes... I know there are cases where this isn't possible, but when it is, it'd be nice to have it.

**Comment by pekkaklarck** [comment]
The problem with including source information is that it only makes sense with files on the machine where spec files (or HTML files) are generated. That said, I see it being a very important feature e.g. with editors. Some initial ideas about this below:

1. This functionality should be configurable with a command line option. When enabled, source would be added to XML spec files but shown also in HTML output files. Not sure what the option should be named, perhaps something like `--includesource` with a suitable short option.

2. Should also configurable programmatically.

3. I don't think it is a good idea to enable this automatically with `*.libspec` files (#3491). It's easy to enable the functionality when needed and documentation is easier if there are no special cases. I'm not strongly against this, though.

4. Source should be stored as an absolute path to avoid breaking it if the spec file is moved.

5. Should check could keywords also get line numbers. I would expect editors to benefit from this.

6. Need to think should we allow giving the source externally as well. If there are not concrete use cases now, this can easily be added later if needed.

**Comment by fabioz** [comment]
I agree about all those points... some observations:

1/2/3. the use case is automatically generating those for libraries that don't have them, so, having the --includesource as an option passed from the client seems good (being default is not needed).

5. Having line numbers would be nice for a go to definition... some notes here:
  - the file for the main imported module should be there globally for the spec (as we want to regenerate the libspec if it changes).
  - there should be a source tied to each specific keyword -- it's not uncommon in Python to have the functions in one place and having the API be just a bunch of imports, so, we want the actual place of the function (it may be skipped if it matches the global filename).

6. I don't think this is needed for now.

**Comment by aaltat** [comment]
With dynamic libraries, like SeleniumLibrary, keywords can locate in multiple files. It could be good option to have keyword specific location instead of library level location.

**Comment by pekkaklarck** [comment]
Thanks for comments @fabioz and @aaltat. We clearly need also keyword level source information. I think it could work like this:

- Library can have source information as discussed earlier.
- Each keyword can have source information in format `path:lineno` where both parts are optional. If `path` is not present, then the library level source can be used instead.

**Comment by pekkaklarck** [comment]
Prototyped this a bit and here are the findings:

1. With Python based libraries using the static or hybrid API we can easily get the actual function/method implementing the keyword. From it we can get the source file with [inspect.getsourcefile](https://docs.python.org/3/library/inspect.html#inspect.getsourcefile) and line number with [inspect.getsourcelines](https://docs.python.org/3/library/inspect.html#inspect.getsourcelines). We can do this only when using Libdoc and when getting source is explicitly requested. That avoids all possible performance problems.

2. With dynamic libraries we cannot get the actual method in a generic manner. This is a problem in general, but it is an especially big problem because SeleniumLibrary utilizes the dynamic API. We probably could special case libraries using the [PythonLibCore](https://github.com/robotframework/PythonLibCore/) like SL does, but that's not a fully generic solution. Alternatively we could add `get_keyword_source` method to the dynamic API and enhance PythonLibCore to support it. Do you have an opinion about this @aaltat?

3. The dynamic API supporting getting keyword source would mean that also JavalibCore could support it. Any comments about that @Hi-Fi?

4. I don't know could we get keyword source info from Java based libraries utilizing the static API. Need to investigate that but I don't consider it a blocker if that's not possible.

5. Not sure what we should do with the remote API. If the dynamic API is enhanced, we could also enhance the remote API but I'm not sure is it worth the effort at this point.

**Comment by aaltat** [comment]
I don't see problems adding new method for dynamic library API. Should not be that difficult to implement in the lib core.

**Comment by pekkaklarck** [comment]
No, it's not hard to add to the APi or to the DynamicCore. Let's do that.

**Comment by fabioz** [comment]
One thing that just crossed my mind is that it should be possible for filenames to be relative too (if we want libraries to generate it when doing a release and distribute it along the sources, filenames need to stored relative to the `.libspec` location).

**Comment by pekkaklarck** [comment]
Should they be relative to the library installation directory or where? Should this be the default behavior?

**Comment by fabioz** [comment]
They should be relative to the location of the .libspec file (my expectation is that library users will generate it on the ci as a part of the release process and the .libspec file will be put in the place where it'll be distributed relative to other resources).

I think it should be the default (as it's what users that want to distribute the sources should do), whereas the absolute path is valid just for the current machine.

**Comment by pekkaklarck** [comment]
How would this work if you install a lib with pip? Should the lib contain the spec or should you download it separately? In the latter case, how can we known what should be the root dir?

What's in general the use case for distributing spec files if Libdoc can be used to generate them as needed?

Perhaps better to move further discussion to Slack and return her when we agree on the details.

**Comment by fabioz** [comment]
I think that if you install with pip it should be distributed along.

As for the use case, I think there are 2 things:

1. Library discoverability
    Right now it's hard to know which are the available libraries as they can appear to be plain Python modules (so, having a .libspec file at the PYTHONPATH would help a lot there -- right now, I just try to create a .libspec after the user imports it -- and I pre-create the bultin libraries -- so, I can't offer suggestions of the libraries to import unless it was already imported once and that file was opened and had a request for code-completion, which isn't ideal).

2. I think that it's a bit wasteful having to regenerate the .libspec at each user machine.
    I did some measurements here: at my machine, it currently takes around 5 seconds to generate the docs for the libraries which are builtin... I'm not sure how long it would take for other libraries, but the main problem is that right now this relates to the first time a library is used (so, the first time it takes around 0.5 seconds more to generate the code completion and it feels quite sluggish in that case -- sure, it's cached afterwards, but that's a first bad impression... usually I'd give a timeout error if a code completion request takes 0.5 seconds).

**Comment by pekkaklarck** [comment]
I can see it would be convenient if libraries had spec files with them, but I'm afraid it's unrealistic to expect that to happen in the near future in larger scale. There are dozens if not hundreds libraries out there, and waiting for all of them to create a new release with specs will take time. Many of the widely used libraries aren't that actively maintained either, so it's possible there's never a new release with a spec file. It will also take time until maintainers would here about this, and until Robot LSP starts to be in wider use they may not find it worth the effort to add the spec even then.

It's strange that it takes 5 seconds to generate spec files for standard libraries on your machine because on my already old laptop that happens in ~1s. I also don't see why even that long time would be a big problem in real life, because after a user has imported a library, it will take some time until they use first keyword from it. This could also made a bit faster by not first generating a spec file but instead importing the library using Robot's API (that's what #3448 is about). All completion info could be read from the model returned by Robot's API and the model could be saved to a spec file as well.

Automatically discovering what libraries are available is an interesting idea. I can see it would be convenient, but definitely not a high priority feature compared e.g. to keyword completion. One way to find libraries would be doing something like `pip list | grep robotframework`, but it doesn't find all packages and some packages it find may not be libraries. We do have `Framework :: Robot Framework :: Library` classier that would be a very good indicator, but I'm afraid not all libraries have it set and I'm not sure does pip support finding installed packages containing it either. It's shown if you use e.g. `pip show --verbose robotframework-seleniumlibrary`, though.

Anyway, in my opinion Robot LSP should work also without spec files being readily available especially in the beginning. Further discussion about that doesn't really belong to this issue, though. That said, I try to make this feature work so that it's possible to use local paths. It will be next on my agenda but not anymore today.

**Comment by pekkaklarck** [comment]
One thing related to this issue and also to #3448 is that the model returned by the `LibraryDocumentation` factory method will contain source and line number information. That's pretty much a prerequisite for being able to write the info to a spec file, but having this info easily available programmatically is convenient on its own right as well.

**Comment by fabioz** [comment]
> I can see it would be convenient if libraries had spec files with them, but I'm afraid it's unrealistic to expect that to happen in the near future in larger scale. There are dozens if not hundreds libraries out there, and waiting for all of them to create a new release with specs will take time. Many of the widely used libraries aren't that actively maintained either, so it's possible there's never a new release with a spec file. It will also take time until maintainers would here about this, and until Robot LSP starts to be in wider use they may not find it worth the effort to add the spec even then.

Oh, sure, I completely understand that, so, LSP can't rely on having it, but I think that if it can give better support for the libraries that do provide it (even if only a few would do it initially) I think it would be nice.

> It's strange that it takes 5 seconds to generate spec files for standard libraries

I think that the main reason is that I'm starting a shell for each new invocation (I'm handling builtin as any other library which could change, so, starting a new shell is a way around not having modules cached), so, it takes around 0.70s to 0.85s for each invocation (which sums to 5 seconds locally for the 12 builtin libraries) -- most of that time is probably spent importing the same `robot` libraries over and over again when starting robotframework... I did some improvements so that the builtins are done in parallel so that it isn't an issue for startup, but generating for a single library still takes around 0.7 (importing directly has the issue of generating a new version when the code changes). 

I can probably work around that in the future by having a special shell just to generate the libspec files (which I could restart when source files change) to get around the slow `robot` startup issue (as I think that's safer than working with messing with `sys.modules` for a reload).

**Comment by pekkaklarck** [comment]
Yeah, `import robot` isn't blazingly fast. Should actually debug is there something especially stupid we do at import time. At least we compile regexps which may not all be used by Libdoc or even by all execution rounds. Lot of the slowness is likely due to there being lot if modules to be imported and all imports requiring disk access.

**Comment by fabioz** [comment]
Yeap... usually the fix in this case is making imports local instead of global (I must say I default to writing imports locally instead of globally due to this as I've been badly bitten due to that already -- even on test cases as I tend to use pytest with xdist and it pre-imports all tests in all nodes prior to starting execution).

Still, this may mean many things have to be reorganized and it's usually a pretty boring thing to do if it wasn't done from the start.

I did a quick profile here and by doing a `python -m robot.libdoc --format XML:HTML BuiltIn c:/temp/builtin.libspec`, the only thing that stands out is calls to `compile` and filesystem access (in Python 3.8 it imports 425 modules for that).

![image](https://user-images.githubusercontent.com/117621/77353263-6e4a5380-6d1f-11ea-9a39-daccf72b2706.png)

**Comment by pekkaklarck** [comment]
Should probably study how much avoiding regexp compilation would save time. I don't think moving global imports to local imports is worth the effort, especially when the resulting code would be pretty ugly and against Python style guides. Most of the imports are anyway needed during normal execution so the benefit in that case would be small. Anyway, this is a digression from the main topic of this issue and we can continue discussion on Slack or elsewhere.

**Comment by pekkaklarck** [comment]
I added the source and line number information to internal objects in 7c8cef848b5862ed07eee49cc5166f9e593c7cfa. The only missing piece on that level is handling dynamic libraries and that requires adding the new `get_keyword_source` method. I submitted a separate issue #3516 to discuss it more.

The main tasks still to do and decide:

- [ ] Implement `get_keyword_source`. Covered by #3516.

- [ ] Add `source` and `lineno` attributes to Libdoc's model objects and read them from imported libraries and resource files. This is easy.

- [ ] Write the source and line number information to spec files. Need to decide the format:
    - New `<source>` element or new `source=` attribute? I like attribute better.
    - Source and line number as a single entry like `lib.py:20` or separately? I think separate is better and would prefer `source="lib.py" lineno="20"`.
    - Omit source and/or line number if they are not available? If not, should we use `null` and `-1` or what values?

- [ ] Should the source path be absolute or relative? If relative, based on what? I guess relative paths are better as they support distributing spec files (but may break if moved locally). Probably should make the path relative to the directory where the spec file is created but could also support a command line option like `--sourceroot` to override it.

**Comment by fabioz** [comment]
> New <source> element or new source= attribute? I like attribute better.

I like having an attribute better too.

> Source and line number as a single entry like lib.py:20 or separately? I think separate is better and would prefer source="lib.py" lineno="20".

Agreed.

> Omit source and/or line number if they are not available? If not, should we use null and -1 or what values?

I think it should be omitted as all this info is optional (as there are cases when it's not available at all).

> Should the source path be absolute or relative? If relative, based on what?

I think they can be relative by default (relative to the `.libspec` file location) and with a flag to generate absolute paths -- I don't think it needs a `--sourceroot` to override the relative location (I can't really see a use-case for it).

Another thing that would be nice and probably straightforward to add is the `mtime` for the source (as seen by `os.path.getmtime`) as when the language server is restarting it can re-check if things changed to re-generate it (while it's running it can usually track modifications, but that's not the case on a restart).

**Comment by aaltat** [comment]
The `os.path.getmtime` sounds reasonable to do, at least for me. As a separate attribute?

**Comment by pekkaklarck** [comment]
Could you clarify the need for mtime? Isn't that something the LSP server should keep track anyway? How would it even work if the spec is not generated locally?

**Comment by fabioz** [comment]
It's really only useful if the spec is generated locally (and only for `.libspec` files that the language server itself creates). 

The idea is knowing if any source file changed to regenerate the `.libspec`... this info needs to be persisted because it's used when the language server is restarted.

If it's kept in that file, the language server server doesn't need to track it independently -- note that this is just to make things a bit easier for the language server -- if it's not there, the language server needs to create an additional mapping just to include the mtime right after the file `.libspec` file is generated (that's the only missing info which needs to be persisted to know whether a `.libspec` file will need to be regenerated).

If others are added as attributes, the mtime should probably be added as an attribute too.

**Comment by pekkaklarck** [comment]
It still feels better for me that editors keep track on mtime themselves and it's not stored to Libdoc's spec files. In addition to "feelings", I have some more concrete reasons:

- There's only pretty narrow use case for this functionality.
- The feature is only useful with fresh locally created files.
- Editors can easily handle this information themselves. Getting the original source information is a totally different thing, but once editors have it they can do a `stat` call themselves.
- In spec files we needed to store mtime with each keyword whereas editors can just store mtimes for each source file. The common case is that all keywords in a library have the same source so Libdoc would need to do much more work than editors.
- Supporting mtime would make `get_keyword_source` in the dynamic API more complicated.

**Comment by pekkaklarck** [comment]
I got implementation of the `get_keyword_source` (#3516) pretty much ready locally. One thing that occurred to me is that we cannot at all control that information that method returns. It's easy to report and error if the return value has an invalid type, but we need to decide what to do if the path doesn't match any file.

Options include:
1. Make it an error. Doesn't feel that good to me. If we later extend this functionality to remote libraries, they cannot that often return values that would be valid paths on the client side.

2. Ignore the value i.e. don't write it to a spec file. Same problem as above but silencing the error makes it harder to see what's the problem.

3. Just write whatever the value is to the spec file and let consumers handle that. I think this is the best solution as consumers need to handle invalid paths anyway. If spec files are downloaded or just moved around in the local machine, paths are likely to wrong as well.

**Comment by fabioz** [comment]
Ok, skipping `mtime` and having the language server do the work there is ok for me.

As for the value, I think it should just write whatever is given and let the client handle it (as you noted, it won't always be valid anyways).

**Comment by pekkaklarck** [comment]
I was thinking to omit the source path with keywords if it's the same as the library source path. Does anyone see downsides with that? It would remove lot of duplicate information from spec files in the common case.

**Comment by pekkaklarck** [comment]
This ought to be implemented by e0efb36ae201ae2dfc76ab59031ec6e75fcd68dc but need to still think did I miss any corner cases. Docs need to be updated anyway. Testing is appreciated anyway!

**Comment by pekkaklarck** [comment]
This functionality was documented in the User Guide by 7212ed333dcb7536dbe7d40dcaf2fcba9ceffda3 and some corner cases fixed by earlier commits. I guess we can consider this issue done.
